packages/web/components/swap-tool/__tests__/swap-tool.spec.tsx

```diff
@@ -0,0 +1,231 @@
+/* eslint-disable import/no-extraneous-dependencies */
+import { Dec, PricePretty } from "@keplr-wallet/unit";
+import {
+  createCallerFactory,
+  DEFAULT_VS_CURRENCY,
+  FilteredPoolsResponse,
+  NumPoolsResponse,
+} from "@osmosis-labs/server";
+import { getAssetFromAssetList } from "@osmosis-labs/utils";
+import { screen, waitFor } from "@testing-library/react";
+import userEvent from "@testing-library/user-event";
+import { rest } from "msw";
+import mockRouter from "next-router-mock";
+
+import { server, trpcMsw } from "~/__tests__/msw";
+import { renderWithProviders } from "~/__tests__/test-utils";
+import { SwapTool } from "~/components/swap-tool";
+import { AssetLists } from "~/config/generated/asset-lists";
+import { ChainList } from "~/config/generated/chain-list";
+import { appRouter } from "~/server/api/root-router";
+
+jest.mock("next/router", () => jest.requireActual("next-router-mock"));
+
+const createCaller = createCallerFactory(appRouter);
+const caller = createCaller({
+  /**
+   * The asset list and chain list used here are snapshots of the
+   * production asset list. In case of discrepancies due to outdated
+   * assets or chains, consider updating these mocks.
+   */
+  assetLists: AssetLists,
+  chainList: ChainList,
+});
+
+beforeEach(() => {
+  server.use(
+    trpcMsw.edge.assets.getUserAssets.query(async (req, res, ctx) => {
+      return res(
+        ctx.status(200),
+        ctx.data(await caller.edge.assets.getUserAssets(req.getInput()))
+      );
+    }),
+    trpcMsw.edge.assets.getAssetPrice.query((_req, res, ctx) => {
+      return res(
+        ctx.status(200),
+        ctx.data(new PricePretty(DEFAULT_VS_CURRENCY, new Dec(1)))
+      );
+    })
+  );
+});
+
+afterEach(() => {
+  mockRouter.setCurrentUrl("/");
+});
+
+it("should return only pool assets when sending a pool id and useOtherCurrencies is true", async () => {
```

Comment:
I'm wondering, should we test this kind of functionality at the UI level or at the function level?

I think tests may be easier to maintain if we test these complex flows at the TS function level, then for UI tests we mock the complex flows from these functions then just test the fact that the UI properly displays it.

I think a good principle for tests is to try to only test one layer at a time, and the other layers can be carefully mocked. This violates that principle as it tests the UI as well as the complex server function.

Reply:
BTW don't let this block the PR

Reply:
Thank you for the review! Overall, this is a solid approach and a good idea for our general testing strategy.

In this case, I wanted to focus on how much confidence the tests give us rather than just their coverage. I feel more assured when we minimize our use of mocks. For the swap tool, in particular, I want us to be very confident that everything will work seamlessly with the client as long as the network provider is up and running. I expect these tests to be infrequent and reserved only for very critical workflows, such as the Swap Tool.